synapse/handlers/federation.py

```diff
@@ -1035,6 +1043,217 @@ async def get_state_ids_for_pdu(self, room_id: str, event_id: str) -> List[str]:
+                        # historical branch. But after the historical branch is
+                        # exhausted and the event comes up again in backfill, we
+                        # will choose the "live" DAG.
+                        if not event_exists_on_remote_server:
+                            found_undiscovered_connected_historical_messages = True
+                            queue.put(
+                                (
+                                    -connected_insertion_event_backfill_item.depth,
+                                    -connected_insertion_event_backfill_item.stream_ordering,
+                                    connected_insertion_event_backfill_item.event_id,
+                                    connected_insertion_event_backfill_item.type,
+                                )
+                            )
+
+                # Second, we need to go and try to find any batch events connected
+                # to a given insertion event (by batch_id). If we find any, we'll
+                # add them to the queue and navigate up the DAG like normal in the
+                # next iteration of the loop.
+                if event_type == EventTypes.MSC2716_INSERTION:
+                    connected_batch_event_backfill_results = (
+                        await self.store.get_connected_batch_event_backfill_results(
+                            event_id, limit - len(event_id_results)
+                        )
+                    )
+                    logger.info(
+                        "get_backfill_events(room_id=%s): connected_batch_event_backfill_results(%s)=%s",
+                        room_id,
+                        event_id,
+                        connected_batch_event_backfill_results,
+                    )
+                    for (
+                        connected_batch_event_backfill_item
+                    ) in connected_batch_event_backfill_results:
+                        if (
+                            connected_batch_event_backfill_item.event_id
+                            not in event_id_results
+                        ):
+                            queue.put(
+                                (
+                                    -connected_batch_event_backfill_item.depth,
+                                    -connected_batch_event_backfill_item.stream_ordering,
+                                    connected_batch_event_backfill_item.event_id,
+                                    connected_batch_event_backfill_item.type,
+                                )
+                            )
+
+            # If we found a historical branch of history off of the message lets
+            # navigate down that in the next iteration of the loop instead of
+            # the normal prev_event chain.
+            if not found_undiscovered_connected_historical_messages:
+                event_id_results.add(event_id)
```

Comment:
New piece here. This is part of the forking choice. If we found an undiscovered historical branch, do that instead. Otherwise, continue on the normal DAG `prev_event` navigation.